pkg/deploy/deploy.go

```diff
@@ -158,20 +159,42 @@ func (d *deployer) deploy(ctx context.Context, rgName, deploymentName, vmssName
 	for i := 0; i < 3; i++ {
 		d.log.Printf("deploying %s", deploymentName)
 		err = d.deployments.CreateOrUpdateAndWait(ctx, rgName, deploymentName, deployment)
-		if serviceErr, ok := err.(*azure.ServiceError); ok &&
-			serviceErr.Code == "DeploymentFailed" &&
-			i < 1 {
-			// on new RP deployments, we get a spurious DeploymentFailed error
-			// from the Microsoft.Insights/metricAlerts resources indicating
-			// that rp-lb can't be found, even though it exists and the
-			// resources correctly have a dependsOn stanza referring to it.
-			// Retry once.
-			d.log.Print(err)
-			continue
+		serviceErr, isServiceError := err.(*azure.ServiceError)
+
+		// Check for two known errors that we know how to handle.
+		if isServiceError && serviceErr.Code == "DeploymentFailed" && len(serviceErr.Details) > 0 {
+			outerCode, outerCodeOk := serviceErr.Details[0]["code"].(string)
+			message, messageOk := serviceErr.Details[0]["message"].(string)
+
+			if outerCodeOk && messageOk && outerCode == "BadRequest" {
+				innerErr := map[string]interface{}{}
+
+				if jsonErr := json.Unmarshal([]byte(message), &innerErr); jsonErr == nil {
+					innerCode, innerCodeOk := innerErr["code"].(string)
+					innerMessage, innerMessageOk := innerErr["message"].(string)
+
+					if i < 1 && innerCodeOk && innerMessageOk && innerCode == "ResourceNotFound" && strings.Contains(innerMessage, "Microsoft.Network/loadBalancers/rp-lb") {
+						// on new RP deployments, we get a spurious DeploymentFailed error
+						// from the Microsoft.Insights/metricAlerts resources indicating
+						// that rp-lb can't be found, even though it exists and the
+						// resources correctly have a dependsOn stanza referring to it.
+						// Retry once.
+						d.log.Print(err)
+						continue
+					} else if innerCodeOk && innerCode == "CannotModifyProbeUsedByVMSS" {
+						// removes the probe reference so we can update the lb rules
+						if retry := d.vmssCleaner.UpdateVMSSProbes(ctx, rgName); retry {
+							continue
+						}
+					}
```

Comment:
why is it needed to add that check again? We are not modifying the probes (at least not yet), and so that effectively isn't ever going to be used until we decide to move the probe to a different port or modify the path (which isn't super likely), until then I am not sure that we want to have unused code (this is why I removed it completely and didn't let that piece of code in the PR that I reverted)

Reply:
Hmm, I might have been missing some of the context on why you added this in originally.

I added it back here because my understanding was that it was an improvement to make the deployment process more robust to errors and didn't have any downsides. But if it's not needed yet or might cause problems, I'll go ahead and take it out and we can reconsider it in another PR.

Reply:
Gotcha, I'll go ahead and remove it from this PR in an upcoming commit then.